ucp/useraccounts_module.php

```diff
@@ -319,57 +323,97 @@ function main($id, $mode)
 						trigger_error($this->language->lang('FORM_INVALID') . '<br><br>' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
 					}
 
-					$delete_sql = empty($markedbadges) ? '' : ' AND ' . $this->db->sql_in_set('badge_id', $markedbadges, true); // true for NOT IN, false (default) for IN
+					// Query user badges
+					$sql = "SELECT badge_id
+						FROM $user_badge_table
+						WHERE user_id = $this_user_id";
+					if (!($result = $this->db->sql_query($sql)))
+					{
+						trigger_error($this->user->lang('UCP_UBADGE_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
+					}
+					$user_badges = $this->db->sql_fetchrowset($result);
+					$this->db->sql_freeresult($result);
 
 					// Delete all user badges that are NOT IN the array of marked badges or all if none is selected.
+					$delete_sql = empty($markedbadges) ? '' : ' AND ' . $this->db->sql_in_set('badge_id', $markedbadges, true); // true for NOT IN, false (default) for IN
 					$sql = 'DELETE FROM ' . $user_badge_table . '
 						WHERE user_id = ' . $this_user_id . "
 						$delete_sql";
-					$this->db->sql_query($sql);
+					if (!$this->db->sql_query($sql))
+					{
+						trigger_error($this->user->lang('UCP_DELETE_UNSELECTED_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
+					}
 
-					foreach ($markedbadges as $markedbadge)
+					if ($markedbadges)
 					{
-						// Reject duplicates.
-						$sql = "SELECT COUNT(*) as duplcount
-							FROM $user_badge_table
-							WHERE badge_id = $markedbadge
-							AND user_id = $this_user_id";
+						// Query valid badges
+						$sql = "SELECT badge_id
+							FROM $badge_table
+							WHERE badge_default = TRUE";
 						if (!($result = $this->db->sql_query($sql)))
 						{
-							trigger_error($this->user->lang('UCP_DUPLICATE_USER_BADGE_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
+							trigger_error($this->user->lang('UCP_DEFAULT_BADGES_QUERY_FAILED') . '<br /><br />' . $this->user->lang('RETURN_UCP', '<a href="' . $this->u_action . '">', '</a>'));
 						}
-						$duplicates = (int) $this->db->sql_fetchfield('duplcount');
+						$default_badges = $this->db->sql_fetchrowset($result);
```

Comment:
On closer inspection it would be better to replace sql_fetchrowset with:
```php
$valid_badges = array();
while ($row = $this->db->sql_fetchrow($result))
{
	$valid_badges[] = $row['badge_id'];
}
```

The check below then simplifies even further to `if (in_array($markedbadge, $valid_badges))`